src/main/java/org/elasticsearch/cluster/action/index/NodeIndexDeletedAction.java

```diff
@@ -64,32 +75,56 @@ public void remove(Listener listener) {
     }
 
     public void nodeIndexDeleted(final ClusterState clusterState, final String index, final String nodeId) throws ElasticsearchException {
-        DiscoveryNodes nodes = clusterState.nodes();
+        final DiscoveryNodes nodes = clusterState.nodes();
         if (nodes.localNodeMaster()) {
-            threadPool.generic().execute(new Runnable() {
+            threadPool.generic().execute(new AbstractRunnable() {
+
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warn("[{}]failed to ack index store deleted for index", t, index);
+                }
+
                 @Override
-                public void run() {
+                protected void doRun() throws Exception {
                     innerNodeIndexDeleted(index, nodeId);
+                    lockIndexAndAck(index, nodes, nodeId, clusterState);
+
                 }
             });
         } else {
             transportService.sendRequest(clusterState.nodes().masterNode(),
                     INDEX_DELETED_ACTION_NAME, new NodeIndexDeletedMessage(index, nodeId), EmptyTransportResponseHandler.INSTANCE_SAME);
+            threadPool.generic().execute(new AbstractRunnable() {
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warn("[{}]failed to ack index store deleted for  index", t, index);
+                }
+
+                @Override
+                protected void doRun() throws Exception {
+                    lockIndexAndAck(index, nodes, nodeId, clusterState);
+                }
+            });
         }
     }
 
-    public void nodeIndexStoreDeleted(final ClusterState clusterState, final String index, final String nodeId) throws ElasticsearchException {
-        DiscoveryNodes nodes = clusterState.nodes();
-        if (nodes.localNodeMaster()) {
-            threadPool.generic().execute(new Runnable() {
-                @Override
-                public void run() {
+    private void lockIndexAndAck(String index, DiscoveryNodes nodes, String nodeId, ClusterState clusterState) throws IOException {
+        try {
+            // we are waiting until we can lock the index / all shards on the node and then we ack the delete of the store to the
+            // master
+            final List<ShardLock> locks = nodeEnv.lockAllForIndex(new Index(index), TimeUnit.MINUTES.toMillis(30));
```

Comment:
can we add to the comment that if we have to wait here, it means that the delete shard on in IndicesClusterService failed because the shard was used *but* the shard will delete it self before release the lock. This means that acquiring the lock means the shard was deleted.